### PR TITLE
[2/N] Rs/vllm quantization - Refactor refactor to support non-uniform via config

### PR DESCRIPTION
**Pull request by robertgshaw2-neuralmagic**
Refactored to support nonuniform quantization by adding a new layer of Abstraction.

Now, `SmoothQuantLinearMethod` can hold a `SmoothQuantFormat`, which implements the details of how to do quant and dequant operations. There are two `SmoothQuantFormat` classes:
- `SmoothQuantDynamicPerToken`
- `SmoothQuantStaticPerTensor`

We have the following lifecycle:
- `LinearMethod` is created during `get_model`, has access to `QuantizationConfig`
- `Layer` is initialized and passed a `LinearMethod`
- `Layer` calls `LinearMethod.create_weights`, which creates a dictionary of weights and metadata
- `Layer` calls `LinearMethod.apply_weights` during inference, passing the dictionary created during `create_weights`

This PR modifies the `LinearMethod.create_weights` API to receive a `layer_name` as argument. The `LinearMethod` then looks in the `config` to determine which `SmoothQuantFormat` to use for the layer with `layer_name`
- As a result, the `LinearMethod` is responsible for parsing the config from disk and making decisions about what the inference format should look like. In this specific case, since the `SmoothQuantConfig` is not very good, we just match on the suffix `qkv` to determine what each layer should use --> but for SparseMLConfig, we could use a similar structure

In this PR, the `SmoothQuantFormat` is passed in the dictionary returned by `create_weights` and then is used by `apply_weights`

### In Summary

I think this is a good overall structure because it:
- (a) allows us to make minimal changes to the existing models
- (b) allows us to make no changes to the model loading lifecycle (i.e. config / constructor / linear method) ** critically requires having one LinearMethod that propagates through the whole model
- (c) encapsulates the nonuniform logic into the `LinearMethod`, allowing us to have a clean interface into 

### For SparseML Models

We could imagine the following architecture:

#### Config
Config is responsible for:
- loading config from disk
- mapping layer_names --> `SparseMLFormat`

```python
class SparseMLConfig
    def from_dict()
    
    def get_layer_format(layer_name):
         return SparseMLFormat
```

#### LinearMethod
Config is responsible for:
- interface between layers and kernels (so LinearMethod is what is used by the model)

```python
class SparseMLLinearMethod:
    def __init__(self, sparseml_config)
          self.sparseml_config = sparseml_config
          
    def create_weights(layer_name, ...):
          # this, e.g. is where nonuniform might be supported
          format = self.sparseml_config.get_layer_format(layer_name)
          
          weights = format.get_weights()
          weights["format"] = format
          
          return weights
     
     # wrapper around the SparseML format
     def apply_weights(x, weights, ...)
           format = weights["format"]
           weights = weights["weights"]
           
           return format.apply_weights(x, weights)
```

#### SparseMLFormat
Format is responsible for:
- actual weight creation and forward

```python
class SparseMLLinearMethod:
    def __init__(self, sparseml_config)
          self.sparseml_config = sparseml_config
          
    def get_weights(sizes):
         # returns dictionary , e.g.
         return {
             "weights": x
             "scales": y
         }
     
     def apply_weights(weights, x):
         # calls cuda kernel 
         return output
```

Sample Formats:
    - `W8A8DynamicPerToken`
    - `SparseW8A8StaticPerTensorAsymmetric`
    - `W4A8DynamicPerToken`
    - ...

### CODE REVIEW

File: vllm/model_executor/layers/quantization/smoothquant/config.py [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+            qkv_idxs = { "q": 0, "k": 1, "v": 2 }
+            if shard_id not in qkv_idxs:
+                raise ValueError(f"Invalid shard_id {shard_id}")
+            shard_id = qkv_idxs[shard_id]
+        elif type(shard_id) != int:
+            raise ValueError(f"Invalid shard id {shard_id}")
+
+        return param[shard_id], loaded_weight
+
+    def get_layer_format(self, layer_name: str) -> SmoothQuantFormat:
```
Comment (robertgshaw2-neuralmagic) [inline vllm/model_executor/layers/quantization/smoothquant/config.py:145]: This should be moved to `SmoothQuantConfig` I think

Comment (robertgshaw2-neuralmagic) [reply vllm/model_executor/layers/quantization/smoothquant/config.py:145]: Or at least the core logic should. This api probably still needs to exist, but this logic fits better conceptually with the spirit of the `QuantizationConfig`

File: vllm/model_executor/layers/linear.py [hunk 2]
```diff
@@ -62,17 +65,20 @@ class UnquantizedLinearMethod(LinearMethodBase):
     def __init__(self, separate_bias_add: bool = False):
         self.separate_bias_add = separate_bias_add
 
-    def create_weights(self, input_size_per_partition: int,
-                       output_size_per_partition: int, input_size: int,
-                       output_size: int,
-                       params_dtype: torch.dtype, logical_widths: Optional[List[int]]) -> Dict[str, Any]:
-        weight = Parameter(torch.empty(output_size_per_partition,
```
[the remark this answers is not in the document]
Comment (varun-sundar-rabindranath) [reply vllm/model_executor/layers/linear.py:69]: I see what is happening.

File: vllm/model_executor/layers/quantization/base_config.py [hunk 3]
```diff
@@ -51,14 +51,12 @@ def get_from_keys(config: Dict[str, Any], keys: List[str]) -> Any:
...
         raise NotImplementedError
 
-    @abstractmethod
     def get_scaled_act_names(self) -> List[str]:
         """Returns the activation function names that should be post-scaled.
-
         For now, this is only used by AWQ.
         """
-        raise NotImplementedError
+        return []
```
Comment (varun-sundar-rabindranath) [inline vllm/model_executor/layers/quantization/base_config.py:62]: Why does this have to be non-abstract ?

**Comment by varun-sundar-rabindranath** [comment]
LGTM 👍

File: vllm/model_executor/layers/quantization/base_config.py (hunk 3 above)
Comment (robertgshaw2-neuralmagic) [reply vllm/model_executor/layers/quantization/base_config.py:62]: I just moved it b/c AWQ is the only one that uses this

File: vllm/model_executor/layers/quantization/smoothquant/config.py [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+        """List of filenames to search for in the model directory."""
+        return [
+            "quant_config.json",
+        ]
+
+    @classmethod
+    def from_config(cls, config: Dict[str, Any]) -> "SmoothQuantConfig":
+        layer_format_map: Dict[str, str] = {}
+        for layer_key, format in config.items():
+            if format in FORMAT_REGISTRY:
```
[the remark this answers is not in the document]
Comment (robertgshaw2-neuralmagic) [reply vllm/model_executor/layers/quantization/smoothquant/config.py:81]: The validation happens in the construtor

**Pull request merged**